Converting Scala Request to Java Request throws exception

Play-Scala 2.7.3

Simple `GET` route:
```
GET     /                           controllers.HomeController.index
```

Action:
```scala
  def index() = Action { implicit request: Request[AnyContent] =>
    request.asJava.body // BOOOM
    ...
  }
```

Exception:
```
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[ClassCastException: play.api.mvc.AnyContentAsEmpty$ cannot be cast to play.mvc.Http$RequestBody]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:351)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:267)
	at play.core.server.AkkaHttpServer$$anonfun$1.applyOrElse(AkkaHttpServer.scala:448)
	at play.core.server.AkkaHttpServer$$anonfun$1.applyOrElse(AkkaHttpServer.scala:446)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:453)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:92)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:92)
Caused by: java.lang.ClassCastException: play.api.mvc.AnyContentAsEmpty$ cannot be cast to play.mvc.Http$RequestBody
	at play.core.j.RequestImpl.body(JavaHelpers.scala:389)
	at controllers.HomeController.$anonfun$index$1(HomeController.scala:11)
	at scala.Function1.$anonfun$andThen$1(Function1.scala:85)
	at play.api.mvc.ActionBuilderImpl.invokeBlock(Action.scala:489)
	at play.api.mvc.ActionBuilderImpl.invokeBlock(Action.scala:487)
	at play.api.mvc.ActionBuilder$$anon$10.apply(Action.scala:426)
	at play.api.mvc.Action.$anonfun$apply$2(Action.scala:98)
	at play.api.libs.streams.StrictAccumulator.$anonfun$mapFuture$4(Accumulator.scala:184)
	at scala.util.Try$.apply(Try.scala:210)
	at play.api.libs.streams.StrictAccumulator.$anonfun$mapFuture$3(Accumulator.scala:184)
```

Problem is that Scala's `AnyContent` can not be mapped to Java's `Http.RequestBody` class.
Also [this line](https://github.com/playframework/playframework/blob/48f2b6cd1e4d2040907505d16f71d67fb2cf582d/core/play/src/main/scala/play/api/mvc/Request.scala#L82) causes a problem because the `case` clause should not be entered, but it is... `[...]` type is ignored.

--- a/core/play/src/main/scala/play/api/mvc/Request.scala
+++ b/core/play/src/main/scala/play/api/mvc/Request.scala
@@ -108,12 +108,22 @@ trait Request[+A] extends RequestHeader {
   override def withoutTransientLang(): Request[A] =
     removeAttr(Messages.Attrs.CurrentLang)
 
-  override def asJava: Http.Request = this match {
-    case req: Request[Http.RequestBody @unchecked] =>
-      // This will preserve the parsed body since it is already using the Java body wrapper
-      new Http.RequestImpl(req)
-    case _ =>
+  /**
+   * Be aware that when converting a Scala request to a Java request that the body
+   * will not be converted automatically to a Java equivalent body. For example:
+   * If the Scala request contains a play.api.mvc.RawBuffer it will not be converted into it's Java equivalent
+   * play.mvc.RawBuffer, or an Scala AnyContentAsEmpty will not be converted into a java.util.Optional.empty()
+   * (which is the Play Java equivalent of an empty body). Therefore helper methods like request.asJava.body().asRaw(),
+   * asJson(), etc. will very likely not work. You can however retrieve any stored body object by using
+   * request.asJava.body().as(classOf[Object]).
+   */
+  override def asJava: Http.Request = this.body match {
+    case null =>
       new Http.RequestImpl(this.withBody(null))
+    case rb: Http.RequestBody => // This will preserve the parsed body since it is already using the Java body wrapper
+      new Http.RequestImpl(this.withBody(rb))
+    case rb =>
+      new Http.RequestImpl(this.withBody(new Http.RequestBody(rb)))
   }
 }
 

--- a/core/play/src/test/scala/play/api/mvc/RequestSpec.scala
+++ b/core/play/src/test/scala/play/api/mvc/RequestSpec.scala
@@ -104,9 +104,41 @@ class RequestSpec extends Specification {
         req.attrs.get(y) must beNone
       }
     }
+    "correctly handle asJava" in {
+      "when body is null" in {
+        dummyRequest(body = null).asJava.body() must_== null
+      }
+      "when body is Java RequestBody that contains String" in {
+        val jbBody        = new RequestBody("hello world")
+        val retrievedBody = dummyRequest(body = jbBody).asJava.body()
+        retrievedBody must_== jbBody
+        retrievedBody.as(classOf[String]) must_== "hello world"
+        retrievedBody.as(classOf[Object]) must_== "hello world"
+        retrievedBody.as(classOf[Integer]) must_== null
+      }
+      "when body is Java RequestBody that contains null" in {
+        val jbBody        = new RequestBody(null)
+        val retrievedBody = dummyRequest(body = jbBody).asJava.body()
+        retrievedBody must_== jbBody
+        retrievedBody.as(classOf[Object]) must_== null
+      }
+      "when body of Scala request is not RequestBody but asJava should convert it into one" in {
+        val jbBody        = AnyContentAsEmpty
+        val retrievedBody = dummyRequest().withBody(AnyContentAsEmpty).asJava.body()
+        retrievedBody.getClass must_== classOf[RequestBody]
+        retrievedBody.as(classOf[AnyContentAsEmpty.type]) must_== AnyContentAsEmpty
+        retrievedBody.as(classOf[Object]) must_== AnyContentAsEmpty
+        retrievedBody.as(classOf[Integer]) must_== null
+      }
+    }
   }
 
-  private def dummyRequest(requestMethod: String = "GET", requestUri: String = "/", headers: Headers = Headers()) = {
+  private def dummyRequest(
+      requestMethod: String = "GET",
+      requestUri: String = "/",
+      headers: Headers = Headers(),
+      body: RequestBody = new RequestBody(null)
+  ): Request[RequestBody] = {
     new DefaultRequestFactory(HttpConfiguration()).createRequest(
       connection = RemoteConnection("", false, None),
       method = "GET",
@@ -114,7 +146,7 @@ class RequestSpec extends Specification {
       version = "",
       headers = headers,
       attrs = TypedMap.empty,
-      new RequestBody(null)
+      body = body,
     )
   }
 }

